### PR TITLE
Fix/27702/5 4/forum notification 2

### PR DESCRIPTION
**Pull request by theodortruffer**
https://mantis.ilias.de/view.php?id=27702

### CODE REVIEW

File: Modules/Forum/classes/class.ilForumNotification.php [hunk 1]
```diff
@@ -277,11 +277,21 @@ public static function getCachedNodeData($ref_id)
...
+            $node_data = $DIC->repositoryTree()->getSubTree(
                 $DIC->repositoryTree()->getNodeData($ref_id),
                 true,
                 'frm'
             );
+            $node_data = array_filter($node_data, function ($forum_node) use ($DIC, $ref_id) {
+                // filter out forum if a grp lies in the path (#0027702)
+                foreach ($DIC->repositoryTree()->getNodePath($forum_node['child'], $ref_id) as $path_node) {
+                    if ($path_node['child'] !== $ref_id && $path_node['type'] == 'grp') {
+                        return false;
```
Comment (playamongthestars) [inline Modules/Forum/classes/class.ilForumNotification.php:289]: it doesn't work. This causes an error if you delete a crs/grp-member. 

Whoops\Exception\ErrorException thrown with message "Invalid argument supplied for foreach()"

Stacktrace:
#15 Whoops\Exception\ErrorException in /var/www/databay_ilias/Modules/Forum/classes/class.ilForumNotification.php:257
#14 ilErrorHandling:handlePreWhoops in /var/www/databay_ilias/Modules/Forum/classes/class.ilForumNotification.php:257
#13 ilForumNotification:checkForumsExistsDelete in /var/www/databay_ilias/Modules/Forum/classes/class.ilForumAppEventListener.php:392
#12 ilForumAppEventListener:handleEvent in /var/www/databay_ilias/Services/EventHandling/classes/class.ilAppEventHandler.php:187
#11 ilAppEventHandler:raise in /var/www/databay_ilias/Services/Membership/classes/class.ilParticipants.php:832
#10 ilParticipants:delete in /var/www/databay_ilias/Services/Membership/classes/class.ilParticipants.php:1046
#9 ilParticipants:deleteParticipants in /var/www/databay_ilias/Services/Membership/classes/class.ilMembershipGUI.php:719
#8 ilMembershipGUI:deleteParticipants in /var/www/databay_ilias/Services/Membership/classes/class.ilMembershipGUI.php:392
#7 ilMembershipGUI:executeCommand in /var/www/databay_ilias/Services/UICore/classes/class.ilCtrl.php:210
#6 ilCtrl:forwardCommand in /var/www/databay_ilias/Modules/Course/classes/class.ilObjCourseGUI.php:2258
#5 ilObjCourseGUI:executeCommand in /var/www/databay_ilias/Services/UICore/classes/class.ilCtrl.php:210
#4 ilCtrl:forwardCommand in /var/www/databay_ilias/Services/Repository/classes/class.ilRepositoryGUI.php:348
#3 ilRepositoryGUI:show in /var/www/databay_ilias/Services/Repository/classes/class.ilRepositoryGUI.php:310
#2 ilRepositoryGUI:executeCommand in /var/www/databay_ilias/Services/UICore/classes/class.ilCtrl.php:210
#1 ilCtrl:forwardCommand in /var/www/databay_ilias/Services/UICore/classes/class.ilCtrl.php:175
#0 ilCtrl:callBaseClass in /var/www/databay_ilias/ilias.php:20


-- GET Data --

ref_id                   : 72
cmdClass                 : ilcoursemembershipgui
cmd                      : post
cmdNode                  : u1:jo:82
baseClass                : ilrepositorygui
fallbackCmd              : confirmDeleteParticipants
rtoken                   : bcf4da0e50b8c1ceb9a39ee75cdd1d0b
lang                     : de
limit                    : 50

File: Modules/Forum/classes/class.ilForumNotification.php [hunk 2]
```diff
@@ -277,11 +277,21 @@ public static function getCachedNodeData($ref_id)
...
+                    }
+                }
+                return true;
```
Comment (playamongthestars) [inline Modules/Forum/classes/class.ilForumNotification.php:292]: it doesn't work. This causes an error if you delete a crs/grp-member. (see above)

**Comment by mjansenDatabay** [comment]
@nmatuschek IMO this PR looks good, I cannot reproduce the issue you described.

File: Modules/Forum/classes/class.ilForumNotification.php [hunk 3]
```diff
@@ -277,7 +277,21 @@ public static function getCachedNodeData($ref_id)
...
-            self::$node_data_cache[$ref_id] = $DIC->repositoryTree()->getChildsByType($ref_id, 'frm');
+            $node_data = $DIC->repositoryTree()->getSubTree(
+                $DIC->repositoryTree()->getNodeData($ref_id),
+                true,
+                'frm'
+            );
+            $node_data = array_filter($node_data, function ($forum_node) use ($DIC, $ref_id) {
+                // filter out forum if a grp lies in the path (#0027702)
+                foreach ($DIC->repositoryTree()->getNodePath($forum_node['child'], $ref_id) as $path_node) {
+                    if ($path_node['child'] !== $ref_id && $path_node['type'] == 'grp') {
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumNotification.php]: @theodortruffer Maybe we should use a `(int)` cast here when using the `!==` operator.

File: Modules/Forum/classes/class.ilForumNotification.php (hunk 2 above)
Comment (mjansenDatabay) [reply Modules/Forum/classes/class.ilForumNotification.php:292]: Hmm, **not** in my tests.

File: Modules/Forum/classes/class.ilForumNotification.php (hunk 1 above)
Comment (theodortruffer) [reply Modules/Forum/classes/class.ilForumNotification.php:289]: Hmm that's weird. I can't think of a reason, why this should not be an array. Also, line 257 is not a foreach loop in my code. Could it be, that you're on a different version?

File: Modules/Forum/classes/class.ilForumNotification.php (hunk 3 above)
Comment (theodortruffer) [reply Modules/Forum/classes/class.ilForumNotification.php]: Good point, I added the cast, and also changed the type comparator to ===

File: Modules/Forum/classes/class.ilForumNotification.php (hunk 2 above)
Comment (theodortruffer) [reply Modules/Forum/classes/class.ilForumNotification.php:292]: Can you reproduce it? If so, what's the exact scenario?

File: Modules/Forum/classes/class.ilForumNotification.php (hunk 1 above)
Comment (mjansenDatabay) [reply Modules/Forum/classes/class.ilForumNotification.php:289]: @theodortruffer I'll discuss this with @nmatuschek  This is most probably NOT an issue.

File: Modules/Forum/classes/class.ilForumNotification.php [hunk 4]
```diff
@@ -285,7 +285,7 @@ public static function getCachedNodeData($ref_id)
             $node_data = array_filter($node_data, function ($forum_node) use ($DIC, $ref_id) {
                 // filter out forum if a grp lies in the path (#0027702)
                 foreach ($DIC->repositoryTree()->getNodePath($forum_node['child'], $ref_id) as $path_node) {
-                    if ($path_node['child'] !== $ref_id && $path_node['type'] == 'grp') {
+                    if ((int) $path_node['child'] !== $ref_id && $path_node['type'] === 'grp') {
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumNotification.php]: Maybe we should `(int)`-cast both operands :-).

`(int) $path_node['child'] !== (int) $ref_id`

Comment (theodortruffer) [reply Modules/Forum/classes/class.ilForumNotification.php]: Oops, yes of course we should :) fixed now

**Pull request merged**